sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java

```diff
@@ -170,77 +170,135 @@ public int readInteger() {
    *  }
    */
   public void readBatch(
```

Comment:
I need more context to review this PR
1. where do we read the column index in `VectorizedColumnReader`?
2. where do we leverage the index to skip pages? In `VectorizedRleValuesReader`?

Reply:
Sorry @cloud-fan, I should've add more context in the PR description. Let me try to add here and copy later to there.

1. The column index filtering is largely implemented in parquet-mr (via classes such as `ColumnIndex` and `ColumnIndexFilter`), and the filtered Parquet pages are returned to Spark through the `ParquetFileReader.readNextFilteredRowGroup` and `ParquetFileReader.getFilteredRecordCount` API. Please see #31393 for the related changes in the vectorized reader path.
2. Spark needs more work to handle mis-aligned Parquet pages returned from parquet-mr side, when there are multiple columns and their type width are different (e.g., int and bigint). For this issue, @lxian already gave a pretty good description in [SPARK-34859](https://issues.apache.org/jira/browse/SPARK-34859). To support the case, Spark needs to leverage the API [`PageReadStore.getRowIndexes`](https://javadoc.io/doc/org.apache.parquet/parquet-column/latest/org/apache/parquet/column/page/PageReadStore.html), which returns the indexes of all rows (note the difference between rows and values: for flat schema there is no difference between the two, but for complex schema they're different) after filtering within a Parquet row group. In addition, because there are gaps between pages, we'll need to know what is the index for the first row in a page, so we can keep consuming values from a page and skip them when they are not in the row indexes. This is provided by the `DataPage.getFirstRowIndex` method.